include/libdnf/common/sack/query.hpp

```diff
@@ -0,0 +1,243 @@
+    public:
+        using LogicError::LogicError;
+        const char * get_domain_name() const noexcept override { return "libdnf::Query"; }
+        const char * get_name() const noexcept override { return "UnsupportedOperation"; }
+        const char * get_description() const noexcept override { return "Unsupported operator"; }
+    };
+
+    using FilterFunctionBool = bool(const T & obj);
+    using FilterFunctionCString = char *(const T & obj);
+    using FilterFunctionInt64 = int64_t(const T & obj);
+    using FilterFunctionString = std::string(const T & obj);
+    using FilterFunctionVectorInt64 = std::vector<int64_t>(const T & obj);
+    using FilterFunctionVectorString = std::vector<std::string>(const T & obj);
+
+    Query() = default;
+    explicit Query(const Set<T> & src_set) : Set<T>::Set(src_set) {}
+    explicit Query(Set<T> && src_set) : Set<T>::Set(std::move(src_set)) {}
+
+    //    std::size_t ifilter(FilterFunctionString * getter, QueryCmp cmp, const std::string & pattern);
+    Query & ifilter(std::string (*getter)(const T &), QueryCmp cmp, const std::string & pattern);
+    Query & ifilter(std::vector<std::string> (*getter)(const T &), QueryCmp cmp, const std::string & pattern);
+    Query & ifilter(std::string (*getter)(const T &), QueryCmp cmp, const std::vector<std::string> & patterns);
+    Query & ifilter(
+        std::vector<std::string> (*getter)(const T &), QueryCmp cmp, const std::vector<std::string> & patterns);
+
+    Query & ifilter(int64_t (*getter)(const T &), QueryCmp cmp, int64_t pattern);
+    Query & ifilter(std::vector<int64_t> (*getter)(const T &), QueryCmp cmp, int64_t pattern);
+    Query & ifilter(int64_t (*getter)(const T &), QueryCmp cmp, const std::vector<int64_t> & patterns);
+    Query & ifilter(std::vector<int64_t> (*getter)(const T &), QueryCmp cmp, const std::vector<int64_t> & patterns);
+
+    Query & ifilter(bool (*getter)(const T &), QueryCmp cmp, bool pattern);
+
+    Query & ifilter(char * (*getter)(const T &), QueryCmp cmp, const std::string & pattern);
+
+    /// Get a single object. Raise an exception if none or multiple objects match the query.
+    const T & get() const {
+        if (get_data().size() == 1) {
+            return *get_data().begin();
+        }
+        throw std::runtime_error("Query must contain exactly one object.");
+    }
+
+    /// List all objects matching the query.
+    const std::set<T> & list() const noexcept { return get_data(); }
+
+    // operators; OR at least
+    // copy()
+    using Set<T>::get_data;
+};
+
+
+template <typename T>
+inline Query<T> & Query<T>::ifilter(
+    Query<T>::FilterFunctionString * getter, QueryCmp cmp, const std::string & pattern) {
+
+    for (auto it = get_data().begin(); it != get_data().end();) {
+        auto value = getter(*it);
+        if (match_string(value, cmp, pattern)) {
+            ++it;
+        } else {
+            it = get_data().erase(it);
+        }
+    }
+    return *this;
+}
```

Comment:
I am sorry, but the search will be probably quite slow.
Iteration over the whole dataset and to compare each element with the search pattern according to the comparison type. Then it removes not matching elements from dataset.
We have to discuss it personally.